lib/dry/transaction/step.rb

```diff
@@ -38,39 +39,62 @@ def with(operation: UNDEFINED, call_args: UNDEFINED)
           operation_name,
           new_operation,
           options,
-          new_call_args,
+          new_call_args
         )
       end
 
-      def call(input)
+      def call(input, next_step = NOOP)
         args = [input] + Array(call_args)
+        continue = once(next_step)
+
+        with_broadcast(args, continue) do
+          step_adapter.call(self, *args, &continue)
+        end
+      end
+
+      def once(next_step)
+        continued = false
+
+        -> (input) do
+          if continued
+            input
+          else
+            continued = true
+            next_step.(input)
+          end
+        end
+      end
+
+      def with_broadcast(args, continue)
         broadcast :step_called, step_name, *args
-        result = step_adapter.call(self, *args)
 
-        result.fmap { |value|
+        result = yield.fmap { |value|
           broadcast :step_succeeded, step_name, *args
           value
         }.or { |value|
           broadcast :step_failed, step_name, *args, value
           Failure(StepFailure.new(self, value))
         }
+
+        continue.(result)
```

Comment:
Do you feel this line might make more sense underneath `step_adapter.call(self, *args, &continue)` inside the `with_broadcast do ...` block above? Feels like this particular line isn't about broadcasting at all...

Reply:
You're right, though we cannot have it inside the block because this would `continue` the execution _before_ broadcasting and I believe we want to emit `step_{succeeded,failed}` before calling the next step.